cupy/linalg/decomposition.py

```diff
@@ -224,64 +224,83 @@ def svd(a, full_matrices=True, compute_uv=True):
     # Remark 2: gesvd only supports jobu = 'A' and jobvt = 'A'
     # Remark 3: gesvd returns matrix U and V^H
     # Remark 4: Remark 2 is removed since cuda 8.0 (new!)
-    n, m = a.shape
-
-    # `a` must be copied because xgesvd destroys the matrix
-    if m >= n:
-        x = a.astype(a_dtype, order='C', copy=True)
-        trans_flag = False
-    else:
-        m, n = a.shape
-        x = a.transpose().astype(a_dtype, order='C', copy=True)
+    
+    n, p = a.shape
+    if p > n:
         trans_flag = True
-    mn = min(m, n)
+        # Perform SVD(X.T) not SVD(X) This reduces complexity from
+        # O(np^2) to O(n^2p) if p > n
+        if a.flags.c_contiguous:
+            X = a.astype(a_dtype, order = 'C', copy = True).transpose()
+        else:
+            X = a.transpose().astype(a_dtype, order = 'F', copy = True)
+        n, p = X.shape
+    else:
+        trans_flag = False
+        X = a.astype(a_dtype, order = 'F', copy = True)
+
+
+    MIN = min(n, p)
 
     if compute_uv:
         if full_matrices:
-            u = cupy.empty((m, m), dtype=a_dtype)
-            vt = cupy.empty((n, n), dtype=a_dtype)
+            U = cupy.empty((n, n), dtype = a_dtype, order = 'F')
```

Comment:
Sorry another comment on the style but can you stick to the following?
```suggestion
            U = cupy.empty((n, n), dtype=a_dtype, order='F')
```